### PR TITLE
Capture authorizations from table

### PR DESCRIPTION
**Pull request by mgascam**
Fixes #

#### Changes proposed in this Pull Request
This PR adds a Capture button to each row of the new authorizations list table. This button will allow merchants to capture an authorization from the `Uncaptured` tab in the `Transactions` page in WP Admin. 

It uses mocked data, since the server endpoints are not available yet. For this reason, it is not yet possible to manually test the complete flow. Only the loading and error states can be tested manually at this moment.

The scope of changes on this PR includes:

- UI components: adds a new `CaptureAuthorizationButton` to the list of non captured authorizations
- State management boilerplate in the front end
- Refactor `capture_terminal_payment` endpoint. The functionality needed to capture a terminal payment and to capture an authorization are very similar. To avoid code duplication, I refactored the endpoint to make it work for both types of payments, passing a new `$is_terminal_payment` parameter.

https://user-images.githubusercontent.com/1553182/188485604-d5de638e-8af2-491e-964b-8416a8ac8a3b.mp4

#### Testing instructions

* Set `displayAuthorizations = true ` [here](https://github.com/Automattic/woocommerce-payments/blob/e9fec5a7a492ecbe59171991867e4f831b9d1296/client/transactions/index.tsx#L24). Don't forget to rebuild statics (i.e. `npm start`)
* Navigate to `WP Admin > Transactions > Uncaptured` tab. You should see a `Capture` button in each row of the table.
* Click on any `Capture` button
* Check that the button transitions to loading state
* Check that it errors eventually and displays an error notice.

-------------------

- [ ] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

File: client/data/authorizations/resolvers.ts [hunk 1]
```diff
@@ -64,6 +72,49 @@ export function* getAuthorizations( query: Query ): any {
...
+	const randomAmount = () => {
+		return 1400 + Math.floor( Math.random() * 5000 );
+	};
+
+	const randomDate = dateI18n(
+		'M j, Y / g:iA',
+		moment.utc( new Date() ).local().toISOString()
+	);
+
+	const randomCaptureDate = dateI18n(
```
Comment (anu-rock) [inline client/data/authorizations/resolvers.ts]: Is the time part necessary here? The design screenshot in #4506 doesn't have it.

Also, when the data is eventually wired up with server, would there be a benefit to have this string in `ISO-8601` format?

**Review by kalessil — CHANGES_REQUESTED** [review]
@mgascam, could you please update the branch and look into CI failures?

File: client/data/authorizations/actions.ts [hunk 2]
```diff
@@ -39,3 +51,59 @@ export function updateAuthorizationsSummary(
...
+
+		authorization = {
+			authorization_id: id,
+			captured: true,
+		};
+
+		yield updateAuthorization( authorization as Authorization );
+
+		// Need to invalidate the resolution so that the components will render again.
+		yield dispatch(
```
Comment (anu-rock) [inline client/data/authorizations/actions.ts:100]: I wonder if there's a side effect of invalidating resolutions for selectors on a page where they are not used? Eg. `getAuthorizations` is used on Transactions page but not on Payment Details page. I have not noticed anything, but wanted to confirm.

File: client/data/authorizations/resolvers.ts (hunk 1 above)
Comment (anu-rock) [inline client/data/authorizations/resolvers.ts]: Although mocked data here will eventually be removed, using static (not random) dates would help in writing predictable tests.

File: client/data/authorizations/selectors.ts [hunk 3]
```diff
@@ -18,24 +28,49 @@ import {
  *
  * @return {Object} The list of authorizations for the given query.
  */
-const getAuthorizationsForQuery = ( state: Record< string, any > ) => {
-	return state.authorizations;
+const getAuthorizationsForQuery = (
+	state: Record< string, any >,
+	query: Query
+) => {
+	const index = getResourceId( query );
```
Comment (anu-rock) [inline client/data/authorizations/selectors.ts:35]: Thinking out loud - using a stringified query string as index makes it difficult to visualize how the state is maintained. Would having a simpler index be an oversimplification of the underlying logic?

File: client/data/authorizations/test/selectors.test.ts [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+		authorizations: {
+			byId: {},
+			[ getResourceId( mockQuery ) ]: {
+				data: mockAuthorizations,
+			},
+			summary: {},
+		},
+	};
+
+	test( 'Returns empty disputes list when disputes list is empty', () => {
```
Comment (anu-rock) [inline client/data/authorizations/test/selectors.test.ts]: We may have a typo here - `disputes` in place of `authorizations`.

File: client/transactions/uncaptured/capture-authorization-button.tsx [hunk 5]
```diff
@@ -0,0 +1,42 @@
+/** @format **/
```
Comment (anu-rock) [inline client/transactions/uncaptured/capture-authorization-button.tsx]: I think this button can be moved to another location for its reuse potential.

File: includes/admin/class-wc-rest-payments-orders-controller.php [hunk 6]
```diff
@@ -137,21 +173,26 @@ public function capture_terminal_payment( $request ) {
 			}
 
 			// Update the order: set the payment method and attach intent attributes.
-			$order->set_payment_method( WC_Payment_Gateway_WCPay::GATEWAY_ID );
-			$order->set_payment_method_title( __( 'WooCommerce In-Person Payments', 'woocommerce-payments' ) );
+			if ( $is_terminal_payment ) {
+				$order->set_payment_method( WC_Payment_Gateway_WCPay::GATEWAY_ID );
+				$order->set_payment_method_title( __( 'WooCommerce In-Person Payments', 'woocommerce-payments' ) );
+			}
 			$intent_id     = $intent->get_id();
```
Comment (anu-rock) [inline includes/admin/class-wc-rest-payments-orders-controller.php]: We can move the declaration for these four variables above and merge the below condition with the one above.

File: client/data/authorizations/test/selectors.test.ts (hunk 4 above)
Comment (mgascam) [reply client/data/authorizations/test/selectors.test.ts]: Oops thanks. Fixed in [d4431d4](https://github.com/Automattic/woocommerce-payments/pull/4676/commits/d4431d4d6c8ae4d367f5ab2a69125d17bdcd4eb3)

File: includes/admin/class-wc-rest-payments-orders-controller.php [hunk 7]
```diff
@@ -168,7 +209,7 @@ public function capture_terminal_payment( $request ) {
 				'id'     => $intent->get_id(),
 			];
 
-			$result = $is_intent_captured ? $result_for_captured_intent : $this->gateway->capture_charge( $order, false );
+			$result = ( $is_intent_captured && $is_terminal_payment ) ? $result_for_captured_intent : $this->gateway->capture_charge( $order, false );
```
Comment (anu-rock) [inline includes/admin/class-wc-rest-payments-orders-controller.php]: The call to `capture_charge` is now further hidden down the ternary chain. An opportunity to refactor this ternary into an `if-else`?

File: tests/unit/admin/test-class-wc-rest-payments-orders-controller.php [hunk 8]
```diff
@@ -534,6 +534,399 @@ public function test_capture_terminal_payment_not_found() {
...
+		$this->assertEquals(
+			[
+				'status' => 'succeeded',
+				'id'     => $this->mock_intent_id,
+			],
+			$response_data
+		);
+
+		$result_order = wc_get_order( $order->get_id() );
+		$this->assertEquals( 'woocommerce_payments', $result_order->get_payment_method() );
```
Comment (anu-rock) [inline tests/unit/admin/test-class-wc-rest-payments-orders-controller.php]: On behalf of @kalessil - `assertSame` :)

File: tests/unit/admin/test-class-wc-rest-payments-orders-controller.php [hunk 9]
```diff
@@ -534,6 +534,399 @@ public function test_capture_terminal_payment_not_found() {
...
+		$this->assertEquals( 'WooCommerce Payments', $result_order->get_payment_method_title() );
+		$this->assertEquals( 'processing', $result_order->get_status() );
+		$url = '/wc/v3/' . ( $this->is_wpcom() ? 'sites/3/' : '' ) . 'payments/readers/receipts/' . $this->mock_intent_id;
+		$this->assertStringEndsWith( $url, $result_order->get_meta( 'receipt_url' ) );
```
Comment (anu-rock) [inline tests/unit/admin/test-class-wc-rest-payments-orders-controller.php]: Is it necessary to assert reader receipt here?

File: tests/unit/admin/test-class-wc-rest-payments-orders-controller.php [hunk 10]
```diff
@@ -534,6 +534,399 @@ public function test_capture_terminal_payment_not_found() {
...
+	}
+
+	public function test_capture_authorization_succeeded_intent() {
```
Comment (anu-rock) [inline tests/unit/admin/test-class-wc-rest-payments-orders-controller.php:594]: Most of the tests for `capture_authorization` are similar to ones for `capture_terminal_payment`. Perhaps we can remove tests with exact same assertions and keep the ones with some difference?

File: includes/admin/class-wc-rest-payments-orders-controller.php [hunk 11]
```diff
@@ -103,7 +117,29 @@ public function register_routes() {
...
+	}
+
+	/**
+	 * Handles payment capture.
+	 *
+	 * @param  WP_REST_Request $request Full data about the request.
+	 * @param  bool            $is_terminal_payment Marks if the request is for a Terminal payment (In Person Payment or Interac at the moment).
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	private function handle_payment_capture( WP_REST_Request $request, bool $is_terminal_payment ) {
```
Comment (kalessil) [inline includes/admin/class-wc-rest-payments-orders-controller.php]: This doesn't look right: `capture_terminal_payment` had a clear focus and goal.
Now it's making even more what is violating the single responsibility we tried to follow there.

How can we use the built-in capture functionality from order actions for the new endpoint?

**Review by kalessil — CHANGES_REQUESTED** [review]

File: includes/admin/class-wc-rest-payments-orders-controller.php (hunk 7 above)
Comment (mgascam) [reply includes/admin/class-wc-rest-payments-orders-controller.php]: Absolutely, thanks!

File: includes/admin/class-wc-rest-payments-orders-controller.php (hunk 6 above)
Comment (mgascam) [reply includes/admin/class-wc-rest-payments-orders-controller.php]: Good point thanks!

File: includes/admin/class-wc-rest-payments-orders-controller.php (hunk 11 above)
Comment (mgascam) [reply includes/admin/class-wc-rest-payments-orders-controller.php]: Thanks for the review @kalessil! 

I [pushed a change](https://github.com/Automattic/woocommerce-payments/pull/4676/commits/c352f2ee9e8f35d16c5bfb5a697aae51ceeff8b1) to have dedicated endpoints for `capture_terminal_payment` and `capture_authorization`.

For context, it turns out that the capturing of a terminal payment and an authorization share the step to prepare the order object before calling the API to capture the charge. To avoid code duplication I tried to factor out those common parts into a new `prepare_order_for_capture`. It accepts a `$is_terminal_payment parameter` since there is one additional step to do for Terminal payments. Please let me know what you think when you have a moment.

>How can we use the built-in capture functionality from order actions for the new endpoint?

The capture of the order from the order actions is triggered by an action hook: `woocommerce_order_action_capture_charge `. This action is handled in [WC_Payment_Gateway_WCPay::capture_charge](https://github.com/Automattic/woocommerce-payments/blob/c352f2ee9e8f35d16c5bfb5a697aae51ceeff8b1/includes/class-wc-payment-gateway-wcpay.php#L374). This is the same method that both [`capture_terminal_payment`](https://github.com/Automattic/woocommerce-payments/blob/c352f2ee9e8f35d16c5bfb5a697aae51ceeff8b1/includes/admin/class-wc-rest-payments-orders-controller.php#L149) and [`capture_authorization` ](https://github.com/Automattic/woocommerce-payments/blob/c352f2ee9e8f35d16c5bfb5a697aae51ceeff8b1/includes/admin/class-wc-rest-payments-orders-controller.php#L209) will call eventually. Does this answer your question or am I missing something?

File: tests/unit/admin/test-class-wc-rest-payments-orders-controller.php (hunk 8 above)
Comment (mgascam) [reply tests/unit/admin/test-class-wc-rest-payments-orders-controller.php]: Thanks! Done in [c352f2e](https://github.com/Automattic/woocommerce-payments/pull/4676/commits/c352f2ee9e8f35d16c5bfb5a697aae51ceeff8b1)

File: client/data/authorizations/resolvers.ts (hunk 1 above)
Comment (mgascam) [reply client/data/authorizations/resolvers.ts]: I removed the mocked data in favor of using fixtures in [eaa5fba](https://github.com/Automattic/woocommerce-payments/pull/4676/commits/eaa5fbab64503b208bfa08223c547e90091d6d25) so I guess this comment no longer applies, please let me know otherwise @anu-rock

File: tests/unit/admin/test-class-wc-rest-payments-orders-controller.php (hunk 9 above)
Comment (mgascam) [reply tests/unit/admin/test-class-wc-rest-payments-orders-controller.php]: You are right, it is not needed since the receipt_url is intended for Terminal payments. Removed in [85d34d2](https://github.com/Automattic/woocommerce-payments/pull/4676/commits/85d34d2bf9646943845321a37e726feb0182b1d9)

File: tests/unit/admin/test-class-wc-rest-payments-orders-controller.php (hunk 10 above)
Comment (mgascam) [reply tests/unit/admin/test-class-wc-rest-payments-orders-controller.php:594]: I agree, they are very similar but I could not find any tests that have the exact same assertions.

File: client/transactions/uncaptured/capture-authorization-button.tsx (hunk 5 above)
Comment (mgascam) [reply client/transactions/uncaptured/capture-authorization-button.tsx]: Indeed, thanks. Done in [63d289f](https://github.com/Automattic/woocommerce-payments/pull/4676/commits/63d289fe6e9e4c89e21d197cf31ad4a1057524ef)

File: client/data/authorizations/actions.ts (hunk 2 above)
Comment (mgascam) [reply client/data/authorizations/actions.ts:100]: That's a good question, to answer it I dig the[ available docs](https://unfoldingneurons.com/2020/wordpress-data-store-properties-resolvers) for the @wordpres/data package. This is my understanding of how this will work, apologies if the explanation is too verbose.

The `invalidateResolutionForStoreSelector` is a 'special' type of action used by the @wordpres/data to signal the resolution state of a selector/resolver pair. It has its own internal reducer to deal with this type of actions. In this case, the `getAuthorizations` selector will be marked as unresolved in the internal state (aka `metadada`). This means that next time this selector is called, @wordpres/data will know that it needs to resolve it first (fetch from API). Please note that, even if the `getAuthorizations` selector is not used in the Payments Details page, it exists in the store's metadata. In the case it was already resolved (i.e the merchant navigated to `Uncaptured` before) it will be marked as unresolved and refreshed next time is accessed. In the case that it was never resolved, it will be marked as such.

For above reasons, I don't think there will be any undesired side effects. I hope this explanation makes sense.

File: client/data/authorizations/selectors.ts (hunk 3 above)
Comment (mgascam) [reply client/data/authorizations/selectors.ts:35]: This will be used when we have pagination in place. If you check the transactions state for instance, you'll see that the state has this shape:

Screen Shot 2022-09-08 at 18 53 41

We will have something similar for authorizations when pagination is implemented. I'd prefer to keep it as it is since we will use it eventually, but I am also happy to change it if you think we should. I wonder what would be a better index in this case?

Comment (anu-rock) [reply client/data/authorizations/selectors.ts:35]: Sounds good, thanks for explaining :) We can keep it like this.

File: client/data/authorizations/resolvers.ts (hunk 1 above)
Comment (anu-rock) [reply client/data/authorizations/resolvers.ts]: It looks quite better now 👍

Now that the resolver returns empty array/object, what would be the best way to capture button functionality with some data?

Comment (mgascam) [reply client/data/authorizations/resolvers.ts]: I'd suggest to import the mocked data from the authorizations fixture. In `client/data/authorizations/resolvers.ts`:

```
+import authorizationsFixture from '../../data/authorizations/test/authorizations.fixture.json';
 
export function* getAuthorizations( query: Query ): Generator< unknown > {
-       yield updateAuthorizations( query, [] );
+       yield updateAuthorizations( query, authorizationsFixture );
}
```

File: includes/admin/class-wc-rest-payments-orders-controller.php (hunk 11 above)
Comment (kalessil) [reply includes/admin/class-wc-rest-payments-orders-controller.php]: > Does this answer your question or am I missing something?

Yes.

> Please let me know what you think when you have a moment.

My main concern remains same: IPP capturing logic (incl. preparing the order for capture) accumulated many edge cases and corrections along the way. IMO the new endpoint should be implemented from scratch and use `\WC_Payment_Gateway_WCPay::capture_charge`.

If hooks allow properly capture payment via admin, the simplest approach for the new endpoint would be have basic validation followed by `\WC_Payment_Gateway_WCPay::capture_charge` call.

Did you try this approach and evaluate if the order meta gets properly populated with this approach?

Comment (mgascam) [reply includes/admin/class-wc-rest-payments-orders-controller.php]: > Did you try this approach and evaluate if the order meta gets properly populated with this approach?

Thanks for the clarification. I'll try your suggestion and will follow up with my findings.

Comment (mgascam) [reply includes/admin/class-wc-rest-payments-orders-controller.php]: @kalessil turns out that you were right. I reverted the changes to the `capture_terminal_payment` endpoint in [373f9d0](https://github.com/Automattic/woocommerce-payments/pull/4676/commits/373f9d0dabcd4a27363808174c40e2d6677b35e3).

> If hooks allow properly capture payment via admin, the simplest approach for the new endpoint would be have basic validation followed by \WC_Payment_Gateway_WCPay::capture_charge call.

I also followed your suggestion in the `capture_authorization` endpoint. I kept the same validations that the `capture_terminal_payment` uses, but omitted the call to `attach_intent_info_to_order` because is not needed according to my tests. Finally, I do call `mark_payment_capture_completed`, instead of `update_order_status_from_intent` to update the status of the order.

File: client/types/authorizations.d.ts [hunk 12]
```diff
@@ -5,12 +5,14 @@ export interface Authorization {
 	authorization_id: string;
 	authorized_on: string;
 	capture_by: string;
+	captured: boolean;
 	order: OrderDetails;
-	risk_level: RiskLevel;
+	risk_level: RiskLevel | string;
```
Comment (kalessil) [inline client/types/authorizations.d.ts:9]: Looks excessive, what is the reason for this change?

File: includes/admin/class-wc-rest-payments-orders-controller.php [hunk 13]
```diff
@@ -199,6 +213,78 @@ public function capture_terminal_payment( $request ) {
...
+				return new WP_Error( 'wcpay_intent_order_mismatch', __( 'The payment cannot be captured', 'woocommerce-payments' ), [ 'status' => 409 ] );
+			}
+			if ( ! in_array( $intent->get_status(), [ 'processing', 'requires_capture', 'succeeded' ], true ) ) {
+				return new WP_Error( 'wcpay_payment_uncapturable', __( 'The payment cannot be captured', 'woocommerce-payments' ), [ 'status' => 409 ] );
+			}
+
+			$result = $this->gateway->capture_charge( $order, false );
+
+			if ( 'succeeded' !== $result['status'] ) {
+				$http_code = $result['http_code'] ?? 502;
```
Comment (kalessil) [inline includes/admin/class-wc-rest-payments-orders-controller.php]: Can we inline the variables' value?

File: includes/admin/class-wc-rest-payments-orders-controller.php [hunk 14]
```diff
@@ -199,6 +213,78 @@ public function capture_terminal_payment( $request ) {
...
+						__( 'Payment capture failed to complete with the following message: %s', 'woocommerce-payments' ),
+						$result['message'] ?? __( 'Unknown error', 'woocommerce-payments' )
+					),
+					[ 'status' => $http_code ]
+				);
+			}
+
+			// Actualize order status.
+			$charge    = $intent->get_charge();
+			$charge_id = $charge ? $charge->get_id() : null;
```
Comment (kalessil) [inline includes/admin/class-wc-rest-payments-orders-controller.php]: Hm, I believe we should have a charge available at that point. 
Also looking into `mark_payment_capture_completed`, it's expected `string`.

Can you explore a little bit deeper this, please?

File: client/transactions/uncaptured/index.tsx [hunk 15]
```diff
@@ -112,7 +120,7 @@ export const AuthorizationsList = (): JSX.Element => {
 
 	const { authorizations, isLoading } = useAuthorizations( getQuery() );
 
-	const getRiskColor = ( risk: RiskLevel ) => {
+	const getRiskColor = ( risk: RiskLevel | string ) => {
```
Comment (kalessil) [inline client/transactions/uncaptured/index.tsx]: Also looks excessive.

File: client/data/authorizations/actions.ts [hunk 16]
```diff
@@ -39,3 +54,58 @@ export function updateAuthorizationsSummary(
...
+
+		let authorization = yield apiFetch( {
+			path: `/wc/v3/payments/orders/${ orderId }/capture_authorization`,
+			method: 'post',
+			data: {
+				payment_intent_id: paymentIntentId,
+			},
+		} );
+
+		authorization = {
```
Comment (kalessil) [inline client/data/authorizations/actions.ts]: We should add a note to remove this once we have live data.

File: includes/admin/class-wc-rest-payments-orders-controller.php [hunk 17]
```diff
@@ -76,6 +76,20 @@ public function register_routes() {
 				],
 			]
 		);
+		register_rest_route(
+			$this->namespace,
+			$this->rest_base . '/(?P<order_id>\w+)/capture_authorization',
```
Comment (kalessil) [inline includes/admin/class-wc-rest-payments-orders-controller.php:81]: Also, it'd be great to have it documented under https://github.com/Automattic/woocommerce-payments/blob/develop/docs/rest-api/source/includes/wp-api-v3/order.md

**Comment by mgascam** [comment]
Closing this PR because it grew too big and we decided to split it. Please check [here](https://github.com/Automattic/woocommerce-payments/pull/4962) for more details.
